Functional refactoring of ValueSource

kaapstorm opened the pull request:
This is an alternative to PR #25865

##### SUMMARY
The objective of this change is to simplify the way administrators configure value sources. 

A value source specifies where MOTECH must get a value, like from a case property, or a form question. Currently there are many subclasses of `ValueSource`. This change, and PR #25865, are both attempts to make it simpler to configure integrations by reducing that complexity.

This PR takes a functional approach. It also goes a little further than the previous PR: It reaches the point where administrators no longer need to specify the doc_type of the `ValueSource` subclass they want to use. The code will use the given keys to determine what the user wants. e.g.

BEFORE:
```json
    {
      "doc_type": "FormQuestionMap",
      "form_question": "/data/abnormal_temperature",
      "value_map": {
        "yes": "05ced69b-0790-4aad-852f-ba31fe82fbd9"
      }
    }
```

AFTER:
```json
    {
      "form_question": "/data/abnormal_temperature",
      "value_map": {
        "yes": "05ced69b-0790-4aad-852f-ba31fe82fbd9"
      }
    }
```

It is clear from the fact that the administrator has given a form question and a value map that they want MOTECH to fetch a value from the form question, and convert the value "yes" to a UUID.

Review by commit. :blowfish:  There are a lot of commits, but most of them are small changes, and many of them just move code around. Best to ignore whitespace for commit "[Break up get_commcare_value()](https://github.com/dimagi/commcare-hq/commit/8370ce7a71536842d4410343a67b38fe5b602166)". (I am particularly interested to hear opinions on this commit. It uses a Python 3 feature not yet used in the codebase. Although not as controversial as the "walrus operator", it can be use improperly, but I hope that the way it is used in this, and the commit that follows it, shows that by naming functions clearly, their meaning can be clear.)

kaapstorm committed:
Drop unnecessary method

kaapstorm committed:
get_commcare_value() function

kaapstorm committed:
ConstantValue._get_commcare_value() should return CommCare value

kaapstorm committed:
Drop ValueSource._get_commcare_value()

kaapstorm committed:
serialize() function

kaapstorm committed:
deserialize() function

kaapstorm committed:
deserialize() function for ColumnMapping

kaapstorm committed:
get_import_value() function

kaapstorm committed:
get_value() function

kaapstorm committed:
DirectionTests

kaapstorm committed:
ValueSource.can_import and ValueSource.can_export

kaapstorm committed:
Drop ValueSource.check_direction()

kaapstorm committed:
get_commcare_value() uses attribute not doc_type

kaapstorm committed:
Move value_map into ValueSource

kaapstorm committed:
Move jsonpath into ValueSource

kaapstorm committed:
Break up get_commcare_value()

(ignore whitespace)

kaapstorm committed:
"doc_type" no longer required

kaapstorm committed:
Stickler

kaapstorm committed:
Merge branch 'master' into nh/motech/func_value_source

kaapstorm committed:
Fix test

millerdev reviewed (commented):
I'd like to schedule a time to review this in person if that works for you. I think I like where it's heading, but the `if/elif` branches (similar to case statements) in `get_commcare_value` and `deserialize` are a little unsettling. It seems like you should be able to push those branches into the various implementations of `ValueSource`, but I think that's what you just undid, so I'm not sure what I'm missing.

kaapstorm commented:
Yes, I'd like to chat through this.

The if/elif statements in the original `get_commcare_value()` function were meant to be temporary. I planned to break them back up into `get_commcare_value()` methods of `CaseProperty`, `FormQuestion` and `ConstantValue`, or use `@singledispatch`. I chose `@singledispatch` because it let me keep with the functional paradigm, and make value sources into data classes.

If we keep `doc_type` as a requirement of a value source definition, then we don't need that `if/elif` statement. I don't like it either. But leaving out `doc_type`, or getting it wrong, has proved to be an easy mistake for administrators to make, so I'd love to be able to drop it. Maybe there is a better way to do that than that clumsy `if/elif` statement.

As for `deserialize()`, I'd like to replace all instances of `ConstantString` with `ConstantValue`. `ConstantString` is never being used to return `None` in `deserialize()`, which is the only difference between it and `ConstantValue`. It needs to be dropped.

kaapstorm committed:
Test dataclasses and schema

kaapstorm committed:
make requirements

kaapstorm committed:
Stickler, and common sense

kaapstorm committed:
Deprecate ConstantString

kaapstorm committed:
ValueSource extends JsonObject

kaapstorm committed:
Use schema to instantiate value source

kaapstorm committed:
Use DictProperty not SchemaProperty

kaapstorm committed:
Ignore doc_type

kaapstorm committed:
Wrap validated data

kaapstorm committed:
Fix last test

Left as separate commit to make it clear that location values sources
must be migrated before this PR can be merged.

kaapstorm commented:
I see there are now THIRTY commits in this PR. :roll_eyes: That's ridiculous.

This PR does meander. Some changes are undone later on. It has some red herrings (like the [Test dataclasses and schema](https://github.com/dimagi/commcare-hq/pull/25961/commits/0d0dcceff6f72bb9a3ef772e2a3ab2376a73ca5a) commit). I am open to closing this PR, doing some aggressive rebasing, and opening a new one. (Before later commits can be deployed, location value sources will need to be migrated. So this PR will probably need to be split into two anyway.)

There are still a few loose ends to be wrapped up here: `as_jsonobject(dict(value_source_definition))` should be moved inside the `get_value()` and `get_import_value()` functions. `can_import()` and `can_export()` can be incorporated into `serialize()` and `deserialize()`.

My objective with these two PRs is to weigh the approach here against the more conservative approach taken in PR #25865:

##### That PR:

Pros | Cons
--- | ---
Subclassing DocumentSchema is familiar to HQ devs. | Not obvious how MRO resolves for subclasses.
PR has a much smaller footprint. | `doc_type` definitions are still required, and not obvious to third-party administrators.

##### This PR:

Pros | Cons
--- | ---
Value source definitions are minimalist. | .
Value source classes are much simpler. | Any changes in this area will result in ugly merge conflicts until this branch is merged or deleted.
Value sources can be implemented using `JsonObject`, *attrs* or *dataclasses*. | `JsonObject` is familiar. *attrs* and *dataclasses* libraries are simpler, but not widely used, and brand new, respectively.
This change should make it easier to migrate configurations off Couch, because the most widely used data structure in MOTECH has been decoupled from *couchdbkit*. | .

kaapstorm committed:
Merge branch 'master' into nh/motech/func_value_source

millerdev commented on corehq/motech/value_source.py:
```diff
@@ -104,6 +105,15 @@ def __eq__(self, other):
         )
 
     @classmethod
+    def get_schema_dict(cls) -> dict:
```
What do you think of accepting `**options` here to facilitate subclasses constructing an augmented schema?

kaapstorm replied:
I'm afraid I don't understand. Obviously subclasses are already adding their own properties to their schema, so I don't think that's what you mean. And currently this method is only used by `as_jsonobject()` ... so I think you mean that `as_jsonobject()` would need to accept `**options` to pass to `get_schema_dict()` ... but would (should?) callers of `as_jsonobject()` be able to determine whether the data they pass is valid or not? (Keeping in mind that the schema doesn't affect the definition of its class.)

If that is what you mean, then I'm worried that gives the caller too much rope. I would imagine that if the schema is rejecting valid values, that's a bug, and not up to the caller to resolve. ... The caller might want to modify the schema for it to transform the values it is passing ... but then why should the caller not transform the values before passing them?

I think I'm missing something.

millerdev commented on corehq/motech/value_source.py:
```diff
@@ -274,21 +324,8 @@ def get_import_value(value_source, external_data):
-    if hasattr(value_source, "case_owner_ancestor_location_field"):
-        return get_case_owner_ancestor_location_field(value_source, case_trigger_info)
-    if hasattr(value_source, "form_user_ancestor_location_field"):
-        get_form_user_ancestor_location_field(value_source, case_trigger_info)
```
🐭 ❤️

millerdev commented on corehq/motech/value_source.py:
```diff
@@ -274,21 +324,8 @@ def get_import_value(value_source, external_data):
 
 @singledispatch
 def get_commcare_value(value_source, case_trigger_info):
```
Why can't this become a polymorphic method `<ValueSource>.get_commcare_value(case_trigger_info)`?

kaapstorm replied:
It could. That's probably the most recognisable, and practical, approach.

`ValueSource` and its subclasses were intended to implement what *Clean Code* calls "Data Transfer Objects" (ch. 6), which sings their praises: "DTOs are very useful structures, especially when communicating with [say, remote APIs]. They often become the first in a series of translation stages that convert raw data ... into objects [like cases or tracked entities]."

... Except that's not really what they are. They could also be seen as "Data Classes", which *Refactoring* calls a "bad smell" (ch. 3). Martin Fowler agrees with you, and reckons I should move `get_commcare_value()` back.

If that's a good idea (and it probably is) I guess it makes sense to move `get_value()` and the rest of the functions back as well.

I'd have to open a new PR, cos then this one's name would be wrong. :upside_down_face:

millerdev commented on corehq/motech/dhis2/events_helpers.py:
```diff
@@ -39,14 +40,16 @@ def _get_program(config, case_trigger_info):
     if config.org_unit_id:
-        org_unit_id = get_value(config.org_unit_id, case_trigger_info)
+        value_source = as_jsonobject(dict(config.org_unit_id))
+        org_unit_id = get_value(value_source, case_trigger_info)
```
This seems to be a common pattern. Could it be refactored into a wrapper, or maybe moved into `get_value` itself?
```py
def get_config_value(config_data, case_trigger_info):
    value_source = as_jsonobject(dict(config_data))
    return get_value(value_source, case_trigger_info)
```

millerdev reviewed (commented):
I like where this is headed. Use of schema is great because it keeps the implementation details of each different value source sub-type in their respective classes.

> I am open to closing this PR, doing some aggressive rebasing, and opening a new one.

Probably a good idea, if nothing else to clean up the history.

millerdev commented:
> Value sources can be implemented using `JsonObject`, **attrs** or ...

❗️ please use `attrs`

`attrs` is 1000x better than `JsonObject` because it removes tons of boilerplate while respecting standard object model conventions. I'm using it everywhere these days.

kaapstorm commented:
I love *attrs* too ... but ... *dataclasses* is basically *attrs* in the standard library. [Useful comparison](https://github.com/ericvsmith/dataclasses/issues/60).

kaapstorm committed:
Fix diagnosis mapping value source

kaapstorm commented on corehq/motech/openmrs/workflow_tasks.py:
```diff
@@ -163,11 +170,12 @@ def _get_start_stop_datetime(self, form_config):
-        }
+        values_for_concept = {}
+        for obs in form_config.openmrs_observations:
+            value_source = as_jsonobject(dict(obs.value))
```
> I'm curious why the dict() is needed.

Urgh:

```python
>>> class Foo(JsonObject):
>>>     bar = DictProperty()
>>>
>>> foo = Foo(bar={"eggs": "ham"})
>>> type(foo.bar)
<class 'jsonobject.containers.JsonDict'>
```

`JsonDict` first has to be cast as `dict` before `Schema` will validate it. :expressionless:

kaapstorm commented on corehq/motech/dhis2/tests/test_events_helpers.py:
```diff
@@ -84,7 +84,7 @@ def test_form_processing(self):
                 'org_unit_id': {
                     'doc_type': 'FormUserAncestorLocationField',
-                    'location_field': LOCATION_DHIS_ID
+                    'form_user_ancestor_location_field': LOCATION_DHIS_ID
```
That is not a bad idea! `Schema` can transform data during validation. The [Wrap validated data](https://github.com/dimagi/commcare-hq/pull/25961/commits/7c379ab09a1311a0f13939332fb48a77689cea95) commit uses `Schema.validate()` instead of `is_valid()` to allow us to do that. So I think we should!

kaapstorm commented on corehq/motech/value_source.py:
```diff
@@ -128,13 +138,24 @@ class CaseProperty(ValueSource):
 
+    @classmethod
+    def get_schema_dict(cls) -> dict:
+        schema = super().get_schema_dict().copy()
```
Of course. Thank you.

kaapstorm committed:
Move functions to methods

kaapstorm committed:
Stickler

kaapstorm committed:
No need to copy schema dict

kaapstorm committed:
fixup! Test dataclasses and schema

kaapstorm committed:
Support all Schema params

kaapstorm committed:
Old and new schemas for location value sources

kaapstorm committed:
isort

kaapstorm committed:
Deprecated classes no longer required

kaapstorm committed:
JsonObject -> attrs

kaapstorm committed:
Rename as_jsonobject

kaapstorm committed:
Move validation to schema

kaapstorm committed:
Use type annotations for attr.ib

kaapstorm committed:
Unused

kaapstorm commented:
Closing this PR. Will rebase, clean up commits, and open a new PR.

kaapstorm commented:
... PR #26044
